servlet/src/test/java/io/undertow/servlet/test/path/MultipleMatchingMappingTestCase.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2021 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package io.undertow.servlet.test.path;
+
+import io.undertow.servlet.api.ServletInfo;
+import io.undertow.servlet.test.util.DeploymentUtils;
+import io.undertow.testutils.DefaultServer;
+import io.undertow.testutils.HttpClientUtils;
+import io.undertow.testutils.TestHttpClient;
+import io.undertow.util.StatusCodes;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import javax.servlet.ServletException;
+import java.io.IOException;
+
+/**
+ * Test cases for <a href="https://issues.redhat.com/browse/UNDERTOW-1844">UNDERTOW-1844</a>.
+ *
+ * @author Carter Kozak
+ */
+@RunWith(DefaultServer.class)
+public class MultipleMatchingMappingTestCase {
+    @BeforeClass
+    public static void setup() throws ServletException {
+        DeploymentUtils.setupServlet(
+                new ServletInfo("path", GetMappingServlet.class)
+                        .addMapping("/path/*")
+                        .addMapping("/*")
+                        .addMapping("*.ext"));
+
+    }
+
+    @Test
+    public void testGetMappingMatchingExtensionAndPrefix() throws IOException {
+        TestHttpClient client = new TestHttpClient();
+        try {
+            HttpGet get = new HttpGet(DefaultServer.getDefaultServerURL() + "/servletContext/path/foo.ext");
+            HttpResponse result = client.execute(get);
+            Assert.assertEquals(StatusCodes.OK, result.getStatusLine().getStatusCode());
+            String response = HttpClientUtils.readResponse(result);
+            Assert.assertEquals("Mapping match:EXTENSION\n" +
+                    "Match value:foo\n" +
+                    "Pattern:*.ext", response);
```

Comment:
Without the bugfix, the match value in this case is empty string, however that's because `path` and `.ext` are the same length. if I used a longer prefix, for example replacing `path` with `abcdef` the result would be match value would report `ab` based on a match portion `/abcdef` with the length of `.ext` stripped from the end.

Reply:
I’m not confident this is correct since it takes into account both prefix and extension matches.
I will verify the specification when I have a moment.

Reply:
Alright, so the [Servlet 4.0 specification](https://javaee.github.io/servlet-spec/downloads/servlet-4.0/servlet-4_0_FINAL.pdf) section 12.1 states:

> 1. The container will try to find an exact match of the path of the request to the path
of the servlet. A successful match selects the servlet.
> 2. The container will recursively try to match the longest path-prefix. This is done
by stepping down the path tree a directory at a time, using the ’/’ character as a
path separator. The longest match determines the servlet selected.
> 3. If the last segment in the URL path contains an extension (e.g. .jsp), the servlet
container will try to match a servlet that handles requests for the extension. An
extension is defined as the part of the last segment after the last ’.’ character.
> 4. If neither of the previous three rules result in a servlet match, the container will
attempt to serve content appropriate for the resource requested. If a "default"
servlet is defined for the application, it will be used. Many containers provide an
implicit default servlet for serving content.

Based on that, we're erroneously classifying this as an EXTENSION match. This is confirmed by the section 12.2.2 examples, second example.

When I have a moment, I'll reproduce the example cases in this test. It's unclear if we'd want to change the behavior beyond fixing the exception in a patch version release because there's some potential to break existing applications, but I defer to your judgement.

Reply:
I've added the examples from the servlet spec as a new test case, and updated extension mapping to only take precedence over the default servlet, not other path or prefix matches.